### PR TITLE
Add swap functionality without fees for market admins

### PR DESCRIPTION
**Pull request by rlfarman**
Adds two new external functions, `swapWithoutFee` and `swapFromSupplierWithoutFee` to the Market contract. These are restricted to the `MARKET_ADMIN_ROLE`.

Some refactoring of shared logic into helper functions and name improvement.

Added tests for purchasing multiple removals using the without fees.

### CODE REVIEW

File: contracts/Market.sol [hunk 1]
```diff
@@ -500,30 +500,12 @@ contract Market is
...
-    _bridgedPolygonNORI.permit({
-      owner: _msgSender(),
-      spender: address(this),
-      value: amount,
-      deadline: deadline,
-      v: v,
-      r: r,
-      s: s
-    });
+    ) = _prepareSwap(certificateAmount, deadline, v, r, s);
```
Comment (jaycenhorton) [inline contracts/Market.sol:508]: Use named params

File: contracts/Market.sol [hunk 2]
```diff
@@ -571,32 +553,118 @@ contract Market is
...
+   *
+   * ##### Requirements:
+   *
+   * - Can only be used when this contract is not paused.
+   * - Can only be used when the caller has the `MARKET_ADMIN_ROLE` role.
+   * @param recipient The address to which the certificate will be issued.
+   * @param amount The total purchase amount in bpNORI. This is the total number of removals being purchased.
+   * @param supplier The only supplier address from which to purchase carbon removals in this transaction.
+   * @param deadline The EIP2612 permit deadline in Unix time.
+   * @param v The recovery identifier for the permit's secp256k1 signature
```
Comment (jaycenhorton) [inline contracts/Market.sol:647]:
```suggestion
   * @param v The recovery identifier for the permit's secp256k1 signature.
```

File: contracts/Market.sol [hunk 3]
```diff
@@ -571,32 +553,118 @@ contract Market is
...
+   * @param r The r value for the permit's secp256k1 signature
```
Comment (jaycenhorton) [inline contracts/Market.sol:648]:
```suggestion
   * @param r The r value for the permit's secp256k1 signature.
```

File: contracts/Market.sol [hunk 4]
```diff
@@ -571,32 +553,118 @@ contract Market is
...
+   * @param s The s value for the permit's secp256k1 signature
```
Comment (jaycenhorton) [inline contracts/Market.sol:649]:
```suggestion
   * @param s The s value for the permit's secp256k1 signature.
```

File: contracts/Market.sol [hunk 5]
```diff
@@ -870,6 +938,177 @@ contract Market is
     });
   }
 
+  /**
+   * @notice Prepares an order to be fulfilled.
+   * @dev This function is responsible for calculating the certificate amount, validating and allocating the supply,
+   * and permitting the bpNori transfer.
+   * @param certificateAmount The total amount for the certificate.
+   * @param deadline The EIP2612 permit deadline in Unix time.
+   * @param v The recovery identifier for the permit's secp256k1 signature
```
Comment (jaycenhorton) [inline contracts/Market.sol:947]:
```suggestion
   * @param v The recovery identifier for the permit's secp256k1 signature.
```

File: contracts/Market.sol [hunk 6]
```diff
@@ -870,6 +938,177 @@ contract Market is
...
+   * @param r The r value for the permit's secp256k1 signature
```
Comment (jaycenhorton) [inline contracts/Market.sol:948]:
```suggestion
   * @param r The r value for the permit's secp256k1 signature.
```

File: contracts/Market.sol [hunk 7]
```diff
@@ -870,6 +938,177 @@ contract Market is
...
+   * @param s The s value for the permit's secp256k1 signature
```
Comment (jaycenhorton) [inline contracts/Market.sol:949]:
```suggestion
   * @param s The s value for the permit's secp256k1 signature.
```

File: contracts/Market.sol [hunk 8]
```diff
@@ -870,6 +938,177 @@ contract Market is
     });
   }
 
+  /**
+   * @notice Prepares an order to be fulfilled.
```
Comment (jaycenhorton) [inline contracts/Market.sol:942]:
```suggestion
   * @notice Prepare an order to be fulfilled.
```

File: contracts/Market.sol [hunk 9]
```diff
@@ -870,6 +938,177 @@ contract Market is
...
+    _validateSupply({
+      certificateAmount: certificateAmount,
+      availableSupply: availableSupply
+    });
+    _validatePrioritySupply({
+      certificateAmount: certificateAmount,
+      availableSupply: availableSupply
+    });
+    (countOfRemovalsAllocated, ids, amounts, suppliers) = _allocateSupply(
+      certificateAmount
```
Comment (jaycenhorton) [inline contracts/Market.sol:980]: Use named param

File: contracts/Market.sol [hunk 10]
```diff
@@ -870,6 +938,177 @@ contract Market is
...
+  }
+
+  /**
+   * @notice Prepares an order from a specific supplier to be fulfilled.
+   * @dev This function is responsible for calculating the certificate amount, validating and allocating the supply
+   * from a specific supplier, and permitting the bpNori transfer.
+   * @param certificateAmount The total amount for the certificate.
+   * @param supplier The only supplier address from which to purchase carbon removals in this transaction.
+   * @param deadline The EIP2612 permit deadline in Unix time.
+   * @param v The recovery identifier for the permit's secp256k1 signature
```
Comment (jaycenhorton) [inline contracts/Market.sol:1001]:
```suggestion
   * @param v The recovery identifier for the permit's secp256k1 signature.
```

File: contracts/Market.sol [hunk 11]
```diff
@@ -870,6 +938,177 @@ contract Market is
...
+   * @param r The r value for the permit's secp256k1 signature
```
Comment (jaycenhorton) [inline contracts/Market.sol:1002]:
```suggestion
   * @param r The r value for the permit's secp256k1 signature.
```

File: contracts/Market.sol [hunk 12]
```diff
@@ -870,6 +938,177 @@ contract Market is
...
+   * @param s The s value for the permit's secp256k1 signature
```
Comment (jaycenhorton) [inline contracts/Market.sol:1003]:
```suggestion
   * @param s The s value for the permit's secp256k1 signature.
```

File: contracts/Market.sol [hunk 13]
```diff
@@ -870,6 +938,177 @@ contract Market is
...
+    });
+    uint8 holdbackPercentage;
+    uint256 restrictedSupplierFee;
+    uint256 unrestrictedSupplierFee;
+    for (uint256 i = 0; i < countOfRemovalsAllocated; i++) {
+      unrestrictedSupplierFee = removalAmounts[i];
+      holdbackPercentage = _removal.getHoldbackPercentage({id: removalIds[i]});
+      if (holdbackPercentage > 0) {
+        restrictedSupplierFee =
+          (unrestrictedSupplierFee * holdbackPercentage) /
```
Comment (jaycenhorton) [inline contracts/Market.sol:1083]: You can use muldiv by openzepelin here to improve gas efficiency

File: contracts/Market.sol [hunk 14]
```diff
@@ -571,32 +553,118 @@ contract Market is
...
     uint256 certificateAmount = this
       .calculateCertificateAmountFromPurchaseTotal({purchaseTotal: amount});
-    _validatePrioritySupply({
+    (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    ) = _prepareSwapFromSupplier(
+        certificateAmount,
```
Comment (jaycenhorton) [inline contracts/Market.sol:562]: Use named params

File: contracts/Market.sol [hunk 15]
```diff
@@ -571,32 +553,118 @@ contract Market is
...
-      owner: _msgSender(),
-      spender: address(this),
-      value: amount,
-      deadline: deadline,
-      v: v,
-      r: r,
-      s: s
+      uint256[] memory amounts,
+      address[] memory suppliers
+    ) = _prepareSwap(amount, deadline, v, r, s);
```
Comment (jaycenhorton) [inline contracts/Market.sol:614]: Use named param

File: contracts/Market.sol [hunk 16]
```diff
@@ -571,32 +553,118 @@ contract Market is
...
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external whenNotPaused onlyRole(MARKET_ADMIN_ROLE) {
+    (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    ) = _prepareSwapFromSupplier(amount, supplier, deadline, v, r, s);
```
Comment (jaycenhorton) [inline contracts/Market.sol:665]: Use named params

File: contracts/Market.sol [hunk 17]
```diff
@@ -870,6 +938,177 @@ contract Market is
...
+   * @param operator The message sender.
+   * @param recipient The recipient of the certificate.
+   * @param countOfRemovalsAllocated The number of distinct removal IDs that are involved in fulfilling this order.
+   * @param ids An array of removal IDs involved in fulfilling this order.
+   * @param amounts An array of amounts being allocated from each corresponding removal token.
+   * @param suppliers An array of suppliers.
+   */
+  function _fulfillOrderWithoutFee(
+    uint256 certificateAmount,
+    address operator,
```
Comment (jaycenhorton) [inline contracts/Market.sol:1060]: Is it possible to rename this to `from`? `operator` typically means the address which is an approved spender of the tokens, but this looks like it's using `from: operator` yet operator == `msg.sender` which is confusing (and potentially a misuse of operator params if it's actually relying on msg.sender being the operator somewhere-- which I didn't see but maybe i didn't look close enough)

File: contracts/Market.sol [hunk 18]
```diff
@@ -870,6 +938,177 @@ contract Market is
...
+    });
+    return (countOfRemovalsAllocated, ids, amounts, suppliers);
+  }
+
+  /**
+   * @notice Fulfill an order without charging the Nori fee.
+   * @dev This function is responsible for paying suppliers, routeing tokens to the RestrictedNORI contract, updating
+   * accounting, and minting the Certificate.
+   * @param certificateAmount The total amount for the certificate.
+   * @param operator The message sender.
```
Comment (jaycenhorton) [inline contracts/Market.sol:1051]:
```suggestion
   * @param operator The address to transfer bpNori from.
```
Or something along those lines

File: contracts/Market.sol [hunk 19]
```diff
@@ -368,27 +368,6 @@ contract Market is
     });
   }
 
-  /**
```
Comment (rlfarman) [inline contracts/Market.sol:371]: Moved this to fix eslint errors

File: contracts/Market.sol [hunk 20]
```diff
@@ -567,33 +546,16 @@ contract Market is
   ) external whenNotPaused {
     uint256 certificateAmount = this
       .calculateCertificateAmountFromPurchaseTotal({purchaseTotal: amount});
-    uint256 availableSupply = _removal.getMarketBalance();
```
Comment (rlfarman) [inline contracts/Market.sol:570]: Moved to `_allocateRemovals`

File: contracts/Market.sol [hunk 21]
```diff
@@ -567,33 +546,16 @@ contract Market is
...
-      certificateAmount: certificateAmount,
-      availableSupply: availableSupply
-    });
     (
       uint256 countOfRemovalsAllocated,
       uint256[] memory ids,
       uint256[] memory amounts,
       address[] memory suppliers
-    ) = _allocateSupply(certificateAmount);
-    _purchasingToken.permit({
```
Comment (rlfarman) [inline contracts/Market.sol:585]: Moved to `_permit`

File: contracts/Market.sol [hunk 22]
```diff
@@ -603,14 +565,14 @@ contract Market is
   }
 
   /**
-   * @notice An overloaded version of `swap` that additionally accepts a supplier address and will exchange IERC20WithPermit
-   * tokens for an ERC721 certificate token and transfers ownership of removal tokens supplied only from the specified
-   * supplier to that certificate. If the specified supplier does not have enough carbon removals for sale to fulfill
-   * the order the transaction will revert.
+   * @notice An overloaded version of `swap` that additionally accepts a supplier address and will exchange
+   * IERC20WithPermit tokens for an ERC721 certificate token and transfers ownership of removal tokens supplied only
```
Comment (rlfarman) [inline contracts/Market.sol:569]: Fixed eslint errors with this docstring

File: contracts/Market.sol [hunk 23]
```diff
@@ -747,8 +797,8 @@ contract Market is
   }
 
   /**
-   * @notice Calculates the total quantity of ERC20 tokens required to make a purchase of the specified `amount` (in tonnes of
```
Comment (rlfarman) [inline contracts/Market.sol:750]: Fixed eslint errors and added @return arg name for this function

File: contracts/Market.sol [hunk 24]
```diff
@@ -869,12 +919,33 @@ contract Market is
     return super.supportsInterface({interfaceId: interfaceId});
   }
 
+  /**
```
Comment (rlfarman) [inline contracts/Market.sol:922]: Moved from above

File: contracts/Market.sol [hunk 25]
```diff
@@ -638,33 +600,129 @@ contract Market is
...
-      spender: address(this),
-      value: amount,
-      deadline: deadline,
-      v: v,
-      r: r,
-      s: s
+  }
+
+  /**
+   * @notice Exchange bpNORI tokens for an ERC721 certificate by transferring ownership of the removals to the
```
Comment (amiecorso) [inline contracts/Market.sol:625]:
```suggestion
   * @notice Exchange ERC20 tokens for an ERC721 certificate by transferring ownership of the removals to the
```

File: contracts/Market.sol [hunk 26]
```diff
@@ -638,33 +600,129 @@ contract Market is
...
+   * certificate, without charging a transaction fee.
+   * @dev See [ERC20Permit](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#ERC20Permit) for more.
+   * The message sender must present a valid permit to this contract to temporarily authorize this market
+   * to transfer the sender's bpNORI to complete the purchase. A certificate is minted in the Certificate contract
```
Comment (amiecorso) [inline contracts/Market.sol:629]:
```suggestion
   * to transfer the sender's ERC20 to complete the purchase. A certificate is minted in the Certificate contract
```

File: contracts/Market.sol [hunk 27]
```diff
@@ -638,33 +600,129 @@ contract Market is
...
+      countOfRemovalsAllocated: countOfRemovalsAllocated,
+      ids: ids,
+      amounts: amounts,
+      suppliers: suppliers
     });
-    _fulfillOrder({
+  }
+
+  /**
+   * @notice An overloaded version of `swap` that additionally accepts a supplier address and will exchange bpNORI
```
Comment (amiecorso) [inline contracts/Market.sol:676]: same notes on the presence of "bpNori" in this docstring

File: contracts/Market.sol [hunk 28]
```diff
@@ -924,18 +1028,194 @@ contract Market is
...
+    });
+    suppliers = new address[](countOfRemovalsAllocated).fill({val: supplier});
+    return (countOfRemovalsAllocated, ids, amounts, suppliers);
+  }
+
+  /**
+   * @notice Permits the transfer of an amount of tokens.
+   * @dev This function is responsible permitting the transfer of ERC20 tokens.
+   * @param amount The total purchase amount in ERC20 tokens. This is the combined total price of the removals being
+   * purchased and the fee paid to Nori.
```
Comment (amiecorso) [inline contracts/Market.sol:1133]:
```suggestion
   * purchased and any fees paid to Nori.
```

File: test/checkout.int.t.sol [hunk 29]
```diff
@@ -13,17 +12,6 @@ abstract contract Checkout is UpgradeableMarket {
...
-    keccak256(
-      "ReceiveRemovalBatch(address,address,uint256,uint256,uint256[],uint256[],address,uint256)"
-    );
-
-  function _deployMockERC20() internal returns (MockERC20Permit) {
-    MockERC20Permit impl = new MockERC20Permit();
-    bytes memory initializer = abi.encodeWithSignature("initialize()");
-    return MockERC20Permit(_deployProxy(address(impl), initializer));
-  }
-
```
Comment (amiecorso) [inline test/checkout.int.t.sol:26]: why is all of this code being deleted?  it looks like my tests that test purchases using an alternate ERC20 are also deleted from this file?

**Review by amiecorso — CHANGES_REQUESTED** [review]
looking good! nice refactor of some of our market functions.  just some comments mainly on docstring stuff and then it looks like some tests were deleted that I'd rather be kept in?

**Review by amiecorso — APPROVED** [review]

**Pull request merged**